### PR TITLE
DevTools: Read context values from context dependencies

### PR DESCRIPTION
**Pull request by eps1lon**
## Summary

Instead of walking up the fiber tree to fill context values from the provider values we now move a cursor along the context dependencies whenever we read context.

This will allow us to read from host specific contexts where we don't have access to the context object e.g. `useFormStatus()`. 

For React versions prior to https://github.com/facebook/react/pull/20890 we still need to walk the fiber tree up to visit context providers since we don't have `memoizedValue` on the context dependency in this case.

## How did you test this change?

- [x] react-debug-tools tests
- [x] inspected components with context hooks (e.g. FunctionalContextConsumerWithUpdates) in the shell
- [x] Backwards compat with 16.8: https://codesandbox.io/p/sandbox/serverless-snowflake-ttrp8v 
- [x] Backwards compat with 17.0: https://codesandbox.io/p/sandbox/17-0-react-devtools-context-inspection-forked-rslrsq
- [x] Backwards compat with 18.0: https://codesandbox.io/p/sandbox/18-0-react-devtools-context-inspection-p8ntzj
- [x] Compat with Canary: https://codesandbox.io/p/sandbox/canary-react-devtools-context-inspection-9y5gkc

### CODE REVIEW

File: packages/react-debug-tools/src/ReactDebugHooks.js
Comment (eps1lon) [inline packages/react-debug-tools/src/ReactDebugHooks.js:158]: Ideally we'd throw an error here since that would be a bug if the context dependencies don't line up with the context reads. However, doing so would break APIs like `ReactDebugTools.inspectHooks` where we don't have access to the current fiber (we have tests that would break if we throw here). Which also means that `ReactDebugTools.inspectHooks` would not have support for `useFormStatus`.

File: packages/react-debug-tools/src/ReactDebugHooks.js
Comment (eps1lon) [inline packages/react-debug-tools/src/ReactDebugHooks.js:159]: This is only for `ReactDebugTools.inspectHooks` where we don't have access to the current fiber (we have tests that would break if we throw just because `currentContextDependency` is null). Which also means that `ReactDebugTools.inspectHooks` would not have support for `useFormStatus`.

Should we leave a comment here highlighting the issues with `ReactDebugTools.inspectHooks`, remove that API entirely or did I misunderstand what the desired implementation is?

**Review by hoxyq — APPROVED** [review]
Thanks! Looks good from the DevTools side, but maybe also worth waiting for a review from @sebmarkbage

**Review by hoxyq — APPROVED** [review]
LGTM! Can you try using `@reactVersion` pragma for some tests in `react-debug-tools`?

If yes, then you can try adding a test cases that will cover compatibility with previous version of React?
This is how we run these tests on CI:
https://github.com/facebook/react/blob/a870b2d5494351d75b68c3d9baf03a52fd40a8ef/.circleci/config.yml#L296-L309

**Comment by eps1lon** [comment]
> If yes, then you can try adding test cases that will cover compatibility with previous versions of React?

Problem is that the inspectedElement-test is broken. I quickly looked at it and as far as I can tell, the whole test setup is broken. It uses the specified React versions for both the inspected tree and what Devtools uses. But Devtools uses its bundled React version so the test should ensure that Devtools uses the React version from the repo, while the inspected React tree uses the specified version. Since the test setup uses the specified React version for both, the test crashes since Devtools uses newer React features from the experimental release channel (e.g. `useCashRefresh`).

We also already have tests for context inspection. They just don't run in the first place: https://github.com/facebook/react/blob/a870b2d5494351d75b68c3d9baf03a52fd40a8ef/packages/react-devtools-shared/src/__tests__/inspectedElement-test.js#L1654-L1656

**Pull request merged**
